Fix commands not being Send / Sync in 0.14

luca-della-vedova opened the pull request:
# Objective

Fixes Commands not being `Send` or `Sync` anymore in 0.14 by implementing `Send` and `Sync` for `RawCommandQueue`.

## Solution

Reference discussion in [discord](https://discord.com/channels/691052431525675048/691052431974465548/1259464518539411570).
It seems that in https://github.com/bevyengine/bevy/pull/13249, when adding a `RawCommandQueue` variant to the `InternalQueue`, the `Send / Sync` traits were not implemented for it, which bubbled up all the way to `Commands` not being `Send / Sync` anymore.
I am not very familiar with the ECS internals so I can't say whether the `RawCommandQueue` is safe to be shared between threads, but I know for sure that before the linked PR `Commands` were indeed `Send` and `Sync` so that PR broke "some workflows" (mandatory [xkcd](https://xkcd.com/1172/)).

## Testing

This PR itself includes a compile test to make sure `Commands` will implement `Send` and `Sync`. The test itself fails without the implementation and succeeds with it.
Furthermore, if I cherry pick the test to a previous release (i.e. 0.13) it indeed succeeds, showing that this is a regression specific to 0.14.

hymm commented:
Can you just implement the traits directly on `Commands`? With some of the fields being `pub(crate)` on `RawCommandQueue`, I'm not sure I'm comfortable saying that all the usages will comply with being `Sync` and it's better to keep it as !Sync.

luca-della-vedova commented:
Thanks for the feedback! [Done](https://github.com/bevyengine/bevy/pull/14392/commits/14f6c98ed594880abc76fddaf7bc4ce17f461dc4), but again not 100% sure about how to justify the `Safety`.
From my understanding all the types in the `Commands` struct other than the `InternalQueue::RawCommandQueue` variant are `Send + Sync` which is why making only that variant `Send + Sync` propagated the trait all the way up to `Commands`.

hymm commented:
> From my understanding all the types in the Commands struct other than the InternalQueue::RawCommandQueue variant are Send + Sync which is why making only that variant Send + Sync propagated the trait all the way up to Commands.

Yeah, so the question is whether `InternalQueue` is Send + Sync. I think it's pretty clear it's send, since all the bytes that it's storing internally are Send. Sync is a little trickier because it's exposing it's internal fields. It's probably Sync, but it's not 100% clear that it is. Some internal API might blow up if we try to access the InternalQueue from multiple threads. We know this currently doesn't happen since `InternalQueue` is `!Sync`. Thus we know that `InternalQueue` is only every being access single threaded. So it might be better to keep it `!Sync`, so we don't accidently break things in the future.

But we can say that `Commands` is clearly `Sync` because all the accesses go through `&mut self`. So that's why I suggested implementing it there.

luca-della-vedova commented:
I addressed the feedback in https://github.com/bevyengine/bevy/pull/14392/commits/14f6c98ed594880abc76fddaf7bc4ce17f461dc4 and made the commands themselves Send / Sync

alice-i-cecile commented:
Good fix, thank you :)
